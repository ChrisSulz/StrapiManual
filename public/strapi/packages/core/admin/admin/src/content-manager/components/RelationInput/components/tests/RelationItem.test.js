import React from 'react';

import { lightTheme, ThemeProvider } from '@strapi/design-system';
import { fireEvent, render, screen } from '@testing-library/react';
import { DndProvider } from 'react-dnd';
import { HTML5Backend } from 'react-dnd-html5-backend';

import { RelationItem } from '../RelationItem';

const setup = ({ endAction, testingDnd = false, ...props }) =>
  render(
    <ThemeProvider theme={lightTheme}>
      <DndProvider backend={HTML5Backend}>
        <RelationItem
          ariaDescribedBy="test"
          iconButtonAriaLabel="Drag"
          canDrag={testingDnd}
          id={0}
          index={0}
          name="test_field"
          endAction={endAction}
          {...props}
        >
          First relation
        </RelationItem>
        {testingDnd ? (
          <RelationItem
            name="test_field"
            ariaDescribedBy="test"
            iconButtonAriaLabel="Drag"
            canDrag={testingDnd}
            id={1}
            index={1}
            endAction={endAction}
            {...props}
          >
            Second relation
          </RelationItem>
        ) : null}
      </DndProvider>
    </ThemeProvider>
  );

describe('Content-Manager || RelationInput || RelationItem', () => {
  it('should render and match snapshot', () => {
    const { container } = setup({});

    expect(container).toMatchSnapshot();
  });

  it('should render endAction and match snapshot', () => {
    const { getByText } = setup({ endAction: <div>end action here</div> });

    expect(getByText('end action here')).toBeInTheDocument();
  });

  describe('Reordering relations', () => {
    it('should not move with arrow keys if the button is not pressed first', () => {
      const updatePositionOfRelationMock = jest.fn();

      setup({
        updatePositionOfRelation: updatePositionOfRelationMock,
        testingDnd: true,
      });

      const [draggedItem] = screen.getAllByText('Drag');

      fireEvent.keyDown(draggedItem, { key: 'ArrowDown', code: 'ArrowDown' });

      expect(updatePositionOfRelationMock).not.toBeCalled();
    });

    it('should move with the arrow keys if the button has been activated first', () => {
      const updatePositionOfRelationMock = jest.fn();

      setup({ updatePositionOfRelation: updatePositionOfRelationMock, testingDnd: true });

      const [draggedItem] = screen.getAllByText('Drag');

      fireEvent.keyDown(draggedItem, { key: ' ', code: 'Space' });
      fireEvent.keyDown(draggedItem, { key: 'ArrowDown', code: 'ArrowDown' });

      expect(updatePositionOfRelationMock).toBeCalledWith(1, 0);
    });

    it('should move with the arrow keys if the button has been activated and then not move after the button has been deactivated', () => {
      const updatePositionOfRelationMock = jest.fn();

      setup({ updatePositionOfRelation: updatePositionOfRelationMock, testingDnd: true });

      const [draggedItem] = screen.getAllByText('Drag');

      fireEvent.keyDown(draggedItem, { key: ' ', code: 'Space' });
      fireEvent.keyDown(draggedItem, { key: 'ArrowDown', code: 'ArrowDown' });
      fireEvent.keyDown(draggedItem, { key: ' ', code: 'Space' });
      fireEvent.keyDown(draggedItem, { key: 'ArrowDown', code: 'ArrowDown' });

      expect(updatePositionOfRelationMock).toBeCalledTimes(1);
    });

    it('should exit drag and drop mode when the escape key is pressed', () => {
      const updatePositionOfRelationMock = jest.fn();

      setup({ updatePositionOfRelation: updatePositionOfRelationMock, testingDnd: true });

      const [draggedItem] = screen.getAllByText('Drag');

      fireEvent.keyDown(draggedItem, { key: ' ', code: 'Space' });
      fireEvent.keyDown(draggedItem, { key: 'Escape', code: 'Escape' });
      fireEvent.keyDown(draggedItem, { key: 'ArrowUp', code: 'ArrowUp' });

      expect(updatePositionOfRelationMock).not.toBeCalled();
    });
  });
});
