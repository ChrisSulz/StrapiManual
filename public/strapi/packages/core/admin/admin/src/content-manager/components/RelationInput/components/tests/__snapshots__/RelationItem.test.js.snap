// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content-Manager || RelationInput || RelationItem should render and match snapshot 1`] = `
.c6 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c0 {
  cursor: default;
}

.c1 {
  background: #ffffff;
  padding-top: 8px;
  padding-right: 16px;
  padding-bottom: 8px;
  padding-left: 16px;
  border-radius: 4px;
  border-color: #dcdce4;
  border: 1px solid #dcdce4;
}

.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  gap: 4px;
}

.c4 {
  width: 100%;
  min-width: 0;
}

.c4 > div[role='button'] {
  cursor: all-scroll;
}

.c5 {
  width: 100%;
  min-width: 0;
}

<div>
  <li
    aria-describedby="test"
    class="c0"
  >
    <div
      class="c1 c2"
      data-handler-id="T0"
    >
      <div
        class="c3 c4"
      >
        <div
          class="c2 c5"
        >
          First relation
        </div>
      </div>
    </div>
  </li>
  <div
    class="c6"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </div>
</div>
`;
